## "서울시 생활이동 데이터 " 전처리 과정
- 해당 데이터를 가공하여 연습문제를 푸는 것이 목표 입니다.
- 다양한 데이터를 통한 반복 연습으로 데이터 전처리에 더욱 능숙해질 수 있도록 하겠습니다.
- EDA를 통해 데이터를 이해하기 위한 전 단계입니다.

### 실습 데이터 안내
- 서울 생활 이동 데이터 : 2022.12.01 ~ 2022.12.31
- 서울 열린데이터 광장 : https://data.seoul.go.kr/dataVisual/seoul/seoulLivingMigration.do
- 자치구 단위 -> 2022년 12월 데이터를 다운로드 받아 진행하면 됩니다.(203MB)

[기본 데이터]
* 생활이동_자치구_2022.12_00시.csv

[데이터 소개]
* 서울 시 구별 이동 2022년 12월 0시~0시 59분 59초 데이터

[변수 소개]
* 도착시간 : 0 -> "0:00~0:59"를 의미 
* 출발 자치구 : 자치구 코드
* 성별 : F(여성), M(남성)
* 나이 : 10 ~ 79세 (5세 단위), 0 ~ 9세, 80세 이상으로 구분
* 유형 : H(야간상주지), W(주간상주지), E(기타)
* 평균 이동 시간 : 동일 열 내 이동인구의 월 단위 평균 이동 시간 (10분단위)
* 이동인구 : 추정 합산값으로 소수점 표출, 개인정보 비식별화를 위해 3명 미만의 경우 * 처리

## 데이터 전처리 실습을 통해 해결해야 할 질문들

- 실습문제 1. 데이터 로딩
 
- 실습문제 2. 기본 정보 확인하기

- 실습문제 3. 결측치 처리하기

- 실습문제 4. 이동유형, 요일, 성별 변수의 value수를 파악하기

- 실습문제 5. 서울 지역의 데이터만 골라보기

- 실습문제 6. 서울지역의 구별로 groupby하여 데이터를 확인하기(유동인구 기준)

- 실습문제 7. '도착시군구코드'를 기준으로 각 구별로 평균 이동시간(분)의 평균과 합을 구하기

- 실습문제 8. '도착시군구코드'를 기준으로 각 구별로 이동인구(합)의 평균과 합을 구하기

- 실습문제 9. 어느 요일의 심야 이동이 제일 많을까요?

- 실습문제 10. 24시간의 데이터를 만들어 실습문제 1~8까지 적용해보겠습니다.(옵션!)

#### 실습문제 1. 데이터 로딩

* Pandas 라이브러리를 활용해서 '생활이동_자치구_2022.12_00시.csv' 데이터를 확인하세요.
    * 아래에 'Pandas'를 Import하고 '생활이동_자치구_2022.12_00시.csv'파일을 'seoul_moving'에 불러오고,  Shape을 확인하는 소스코드를 작성하고 실행해보세요.
    * 데이터 파일 로딩시 참고 사항 
        * 구분자(sep)는 ',' 입니다
        * CP949 인코더를 사용해 주세요

In [1]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
import pandas as pd

seoul_moving = pd.read_csv('생활이동_자치구_2022.12_00시.csv', sep=",", encoding = 'cp949')

#### 실습문제 2. 기본 정보 확인하기
* 'seoul_moving' 데이터의 정보를 확인해보세요.
* 'describe', 'info', 'head' 등 전부 활용해 보겠습니다.

In [2]:
# 데이터 프레임의 Shape을 확인합니다.
seoul_moving.shape

(803552, 10)

In [3]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving.head()

대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분) 이동인구(합)
0  202212  일     0      11010      11010  F   0   EE           27   83.82
1  202212  일     0      11010      11010  F  10   EH           75   17.68
2  202212  일     0      11010      11010  F  10   HH           59    11.8
3  202212  일     0      11010      11010  F  10   WW           66       6
4  202212  일     0      11010      11010  F  10   EE           96   11.92

In [4]:
# 0번 row의 의미
# 2022년 12월 일요일에 종로구(11010)에서 출발하여 0시 ~ 0시59분 사이에 
# 종로구(11010)로 이동한(기타E 에서 기타E로 이동) 0~9세 여성은 총 83.82명이고 
# 이들의 평균 이동시간은 27분이다

In [5]:
# 2번 row의 의미
# 2022년 12월 일요일에 종로구(11010)에서 출발하여 0시 ~ 0시 59분 사이에 
# 종로구(11010)로 이동한(야간상주지H 에서 야간상주지H로 이동) 10~14세 여성은 11.8명이고
# 이들의 평균 이동시간은 59분이다

In [6]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 803552 entries, 0 to 803551
Data columns (total 10 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   대상연월         803552 non-null  int64 
 1   요일           803552 non-null  object
 2   도착시간         803552 non-null  int64 
 3   출발 시군구 코드    803552 non-null  int64 
 4   도착 시군구 코드    803552 non-null  int64 
 5   성별           803552 non-null  object
 6   나이           803552 non-null  int64 
 7   이동유형         803552 non-null  object
 8   평균 이동 시간(분)  803552 non-null  int64 
 9   이동인구(합)      803552 non-null  object
dtypes: int64(6), object(4)
memory usage: 61.3+ MB


#### 실습문제 3. 결측치 처리 ( NaN이 아닌 결측치를 처리 해봅시다)

* '인구이동(합)' column은 이동 인구의 합산인데 object 타입입니다. 

    - 왜 일까요? 

    - Tip! (이동인구 : 추정 합산값으로 소수점 표출, 개인정보 비식별화를 위해 3명 미만의 경우 * 처리)

    - 해당 값을 나름의 기준으로 변경하세요

    - '인구이동(합)' 컬럼의 데이터 타입을 수치형으로 변경하세요 

In [7]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving['이동인구(합)'].value_counts()

*         152200
3.3        12790
6.05       11506
3.29        8981
3.01        7381
           ...  
118.77         1
80.5           1
527.39         1
145.55         1
5.18           1
Name: 이동인구(합), Length: 20277, dtype: int64

In [8]:
# 최빈값과 평균값을 알 수 없는 상황이라 '중위값'으로 대체
seoul_moving.replace("*","1.5",inplace=True)

In [9]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving['이동인구(합)'].value_counts()

1.5       152200
3.3        12790
6.05       11506
3.29        8981
3.01        7381
           ...  
118.77         1
80.5           1
527.39         1
145.55         1
5.18           1
Name: 이동인구(합), Length: 20277, dtype: int64

In [10]:
# 이동인구(합) column의 데이터 타입을 바꿔주세요
# astype float
seoul_moving = seoul_moving.astype({'이동인구(합)':'float'})


In [11]:
seoul_moving.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 803552 entries, 0 to 803551
Data columns (total 10 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   대상연월         803552 non-null  int64  
 1   요일           803552 non-null  object 
 2   도착시간         803552 non-null  int64  
 3   출발 시군구 코드    803552 non-null  int64  
 4   도착 시군구 코드    803552 non-null  int64  
 5   성별           803552 non-null  object 
 6   나이           803552 non-null  int64  
 7   이동유형         803552 non-null  object 
 8   평균 이동 시간(분)  803552 non-null  int64  
 9   이동인구(합)      803552 non-null  float64
dtypes: float64(1), int64(6), object(3)
memory usage: 61.3+ MB


#### 실습문제 4. 이동유형, 요일, 성별 변수의 value수를 파악

* 역시나 value_counts()를 활용해 봅시다.

In [12]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving['이동유형'].value_counts()

EH    292483
WH    153877
EE    133269
HE     64823
WE     54291
EW     46051
HH     22812
WW     19071
HW     16875
Name: 이동유형, dtype: int64

In [13]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving['요일'].value_counts()

토    145738
금    123103
일    119001
목    118500
수    104548
화     98750
월     93912
Name: 요일, dtype: int64

In [14]:
# 12월 요일 수 
# 일 : 4, 월 : 4, 화: 4, 수 : 4, 목 : 5, 금 : 5,토 : 5

In [15]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving['성별'].value_counts()


M    469049
F    334503
Name: 성별, dtype: int64

#### 실습문제 5. 서울 지역의 데이터만 골라 보도록 하겠습니다.

* 우리는 서울시의 데이터만 보겠습니다.

* 출발과 도착이 모두 서울인 데이터만 추출해 주세요.

In [16]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving


대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분)  \
0       202212  일     0      11010      11010  F   0   EE           27   
1       202212  일     0      11010      11010  F  10   EH           75   
2       202212  일     0      11010      11010  F  10   HH           59   
3       202212  일     0      11010      11010  F  10   WW           66   
4       202212  일     0      11010      11010  F  10   EE           96   
...        ... ..   ...        ...        ... ..  ..  ...          ...   
803547  202212  토     0      39000      11250  M  40   EH          213   
803548  202212  토     0      39000      11250  M  45   EH          241   
803549  202212  토     0      39000      11250  M  50   EH          206   
803550  202212  토     0      39000      11250  M  65   EH          169   
803551  202212  토     0      39000      11250  M  70   EE          170   

        이동인구(합)  
0         83.82  
1         17.68  
2         11.80  
3          6.00  
4         11.92  
...         ...  
803547     5.98  
803548     9.50  
803549     3.27  
803550     4.06  
803551     4.67  

[803552 rows x 10 columns]

In [17]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
df = seoul_moving[(seoul_moving['출발 시군구 코드'] < 20000) & (seoul_moving['도착 시군구 코드'] < 20000)]


In [18]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
df


대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분)  \
0       202212  일     0      11010      11010  F   0   EE           27   
1       202212  일     0      11010      11010  F  10   EH           75   
2       202212  일     0      11010      11010  F  10   HH           59   
3       202212  일     0      11010      11010  F  10   WW           66   
4       202212  일     0      11010      11010  F  10   EE           96   
...        ... ..   ...        ...        ... ..  ..  ...          ...   
764447  202212  토     0      11250      11250  M  80   EH          205   
764448  202212  토     0      11250      11250  M  80   WH           49   
764449  202212  토     0      11250      11250  M  80   HE           72   
764450  202212  토     0      11250      11250  M  80   WW           43   
764451  202212  토     0      11250      11250  M  80   EE           37   

        이동인구(합)  
0         83.82  
1         17.68  
2         11.80  
3          6.00  
4         11.92  
...         ...  
764447    66.22  
764448    28.39  
764449    18.92  
764450    28.61  
764451    28.37  

[318756 rows x 10 columns]

#### 실습문제 6. 서울지역의 구 별로 groupby하여 데이터를 확인하겠습니다 (유동인구 기준)

* 유출이 제일 많은 구는?

* 유입이 제일 많은 구는?

* 유출이 제일 적은 구는?

* 유입이 제일 적은 구는?

* 참고자료 서울 구별 코드
    * 11010	종로구
    * 11020	중구
    * 11030	용산구
    * 11040	성동구
    * 11050	광진구
    * 11060	동대문구
    * 11070	중랑구
    * 11080	성북구
    * 11090	강북구
    * 11100	도봉구
    * 11110	노원구
    * 11120	은평구
    * 11130	서대문구
    * 11140	마포구
    * 11150	양천구
    * 11160	강서구
    * 11170	구로구
    * 11180	금천구
    * 11190	영등포구
    * 11200	동작구
    * 11210	관악구
    * 11220	서초구
    * 11230	강남구
    * 11240	송파구
    * 11250	강동구

In [19]:
# 유출이 가장 많은 구와 적은 구
df.groupby(by=['출발 시군구 코드'], as_index= False)['이동인구(합)'].sum().sort_values(by='이동인구(합)', ascending =False)

출발 시군구 코드    이동인구(합)
22      11230  914043.94
13      11140  544488.60
23      11240  537856.34
21      11220  442019.00
1       11020  384586.29
18      11190  367031.71
0       11010  344678.63
4       11050  326683.76
20      11210  317195.53
15      11160  299161.80
24      11250  277088.19
3       11040  272750.27
7       11080  260273.02
2       11030  254949.68
10      11110  251089.42
5       11060  246600.76
12      11130  243815.41
19      11200  240168.98
16      11170  211390.55
11      11120  192602.49
8       11090  179050.65
6       11070  178196.39
14      11150  176714.33
17      11180  111635.17
9       11100  110545.89

In [20]:
# 유입이 가장 많은 구와 적은 구
df.groupby(by=['도착 시군구 코드'], as_index= False)['이동인구(합)'].sum().sort_values(by='이동인구(합)', ascending =False)

도착 시군구 코드    이동인구(합)
22      11230  560148.45
23      11240  545654.27
20      11210  426611.06
13      11140  405455.02
15      11160  379165.25
24      11250  362938.46
21      11220  348472.68
7       11080  344483.49
4       11050  338037.37
10      11110  337764.31
18      11190  308492.86
19      11200  307598.72
5       11060  306710.62
11      11120  299335.70
12      11130  284282.68
3       11040  264444.60
6       11070  253326.29
14      11150  236980.93
16      11170  235797.73
8       11090  222694.54
2       11030  210944.33
1       11020  207531.19
0       11010  182971.69
9       11100  177469.69
17      11180  137304.87

#### 실습문제 7. '도착시군구코드'를 기준으로 각 구별로 평균 이동시간(분)의 평균과 합을 구하기

* groupby 사용, as_index=False로 설정해주세요

In [21]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving_sum = df.groupby(by=['도착 시군구 코드'], as_index= False)[['평균 이동 시간(분)','이동인구(합)']].sum().sort_values(by='도착 시군구 코드', ascending =False)

In [22]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving_mean = df.groupby(by=['도착 시군구 코드'], as_index= False)[['평균 이동 시간(분)','이동인구(합)']].mean().sort_values(by='도착 시군구 코드', ascending =False)

In [23]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving_sum.rename(columns={"평균 이동 시간(분)":'총 이동 시간', '이동인구(합)':'총 이동 인구'},inplace=True)

In [24]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving_mean.rename(columns={"평균 이동 시간(분)":'평균 이동 시간', '이동인구(합)':'평균 이동 인구'},inplace=True)

In [25]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
df_seoul_moving = pd.merge(seoul_moving_sum,seoul_moving_mean, how = 'inner', on = '도착 시군구 코드')
df_seoul_moving

도착 시군구 코드  총 이동 시간    총 이동 인구   평균 이동 시간   평균 이동 인구
0       11250  1018180  362938.46  85.626104  30.522114
1       11240  1121744  545654.27  77.148831  37.527804
2       11230  1169335  560148.45  72.234680  34.602696
3       11220   939757  348472.68  66.958105  24.828834
4       11210  1016506  426611.06  74.688170  31.345412
5       11200   942487  307598.72  71.590353  23.364886
6       11190   950718  308492.86  68.703425  22.293168
7       11180   643993  137304.87  73.996668  15.776729
8       11170   896047  235797.73  76.559040  20.146764
9       11160  1014807  379165.25  77.810689  29.072631
10      11150   801305  236980.93  71.885261  21.259615
11      11140  1037579  405455.02  70.012078  27.358638
12      11130   886430  284282.68  72.592744  23.280868
13      11120   913601  299335.70  76.356122  25.017610
14      11110  1039610  337764.31  77.053810  25.034414
15      11100   801371  177469.69  76.825904  17.013679
16      11090   869958  222694.54  73.875510  18.910881
17      11080   972511  344483.49  69.425400  24.591911
18      11070   871440  253326.29  72.246725  21.002014
19      11060   906606  306710.62  67.146052  22.715940
20      11050   890076  338037.37  67.804982  25.751304
21      11040   801431  264444.60  62.203586  20.525039
22      11030   759399  210944.33  63.911715  17.753268
23      11020   784583  207531.19  61.666509  16.311498
24      11010   747473  182971.69  62.237552  15.234945

#### 실습문제 9. 어느 요일의 심야 이동이 가장 많을까요?
* 심야 기준 : 00시~05시

In [26]:
# 0:00~ 4:59 데이터 불러오기 

seoul_moving_0 = pd.read_csv('생활이동_자치구_2022.12_00시.csv', sep=",", encoding = 'cp949')
seoul_moving_1 = pd.read_csv('생활이동_자치구_2022.12_01시.csv', sep=",", encoding = 'cp949')
seoul_moving_2 = pd.read_csv('생활이동_자치구_2022.12_02시.csv', sep=",", encoding = 'cp949')
seoul_moving_3 = pd.read_csv('생활이동_자치구_2022.12_03시.csv', sep=",", encoding = 'cp949')
seoul_moving_4 = pd.read_csv('생활이동_자치구_2022.12_04시.csv', sep=",", encoding = 'cp949')

In [27]:
# 0:00~ 4:59 데이터 합치기 
seoul_night = pd.concat([seoul_moving_0,seoul_moving_1,seoul_moving_3,seoul_moving_4], axis = 0)
seoul_night

대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분) 이동인구(합)
0       202212  일     0      11010      11010  F   0   EE           27   83.82
1       202212  일     0      11010      11010  F  10   EH           75   17.68
2       202212  일     0      11010      11010  F  10   HH           59    11.8
3       202212  일     0      11010      11010  F  10   WW           66       6
4       202212  일     0      11010      11010  F  10   EE           96   11.92
...        ... ..   ...        ...        ... ..  ..  ...          ...     ...
337130  202212  토     4      38000      11240  F  50   HE          272    3.15
337131  202212  토     4      38000      11240  M  30   WE          392       *
337132  202212  토     4      38000      11250  F  55   HE          298    3.27
337133  202212  토     4      39000      11150  F  45   EE          470    3.33
337134  202212  토     4      39000      11210  F  35   EE          489    3.01

[2035581 rows x 10 columns]

In [28]:
# 결측치 처리하기
seoul_night.replace("*","1.5", inplace=True)

In [30]:
# 데이터 타입 변경
seoul_night = seoul_night.astype({'이동인구(합)':'float'})
seoul_night.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2035581 entries, 0 to 337134
Data columns (total 10 columns):
 #   Column       Dtype  
---  ------       -----  
 0   대상연월         int64  
 1   요일           object 
 2   도착시간         int64  
 3   출발 시군구 코드    int64  
 4   도착 시군구 코드    int64  
 5   성별           object 
 6   나이           int64  
 7   이동유형         object 
 8   평균 이동 시간(분)  int64  
 9   이동인구(합)      float64
dtypes: float64(1), int64(6), object(3)
memory usage: 170.8+ MB


In [31]:
# 서울 지역의 데이터만 골라보기
seoul_night = seoul_night[(seoul_night["출발 시군구 코드"] < 20000) & (seoul_night["도착 시군구 코드"] < 20000)]
seoul_night

대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분)  \
0       202212  일     0      11010      11010  F   0   EE           27   
1       202212  일     0      11010      11010  F  10   EH           75   
2       202212  일     0      11010      11010  F  10   HH           59   
3       202212  일     0      11010      11010  F  10   WW           66   
4       202212  일     0      11010      11010  F  10   EE           96   
...        ... ..   ...        ...        ... ..  ..  ...          ...   
323898  202212  토     4      11250      11250  M  80   EH           22   
323899  202212  토     4      11250      11250  M  80   WH           76   
323900  202212  토     4      11250      11250  M  80   WE           49   
323901  202212  토     4      11250      11250  M  80   EE           28   
323902  202212  토     4      11250      11250  M  80   HW           54   

        이동인구(합)  
0         83.82  
1         17.68  
2         11.80  
3          6.00  
4         11.92  
...         ...  
323898    18.97  
323899    18.92  
323900     9.52  
323901     9.52  
323902     9.52  

[943097 rows x 10 columns]

In [32]:
# 심야 시간대 어느 요일의 이동이 가장 많을까?
seoul_moving_mean = seoul_night.groupby(by=['요일'], as_index= False)['이동인구(합)'].mean().sort_values(by='이동인구(합)', ascending =False)

In [33]:
seoul_moving_mean

요일    이동인구(합)
5  토  21.684090
0  금  19.610873
1  목  19.346606
4  일  18.074044
2  수  16.648447
6  화  15.740621
3  월  14.468381

#### 실습문제 10. 24시간의 데이터를 만들어 실습문제 1~8까지 적용해보겠습니다.(옵션!)


In [34]:
seoul_moving_0 = pd.read_csv('생활이동_자치구_2022.12_00시.csv', sep=",", encoding = 'cp949')
seoul_moving_1 = pd.read_csv('생활이동_자치구_2022.12_01시.csv', sep=",", encoding = 'cp949')
seoul_moving_2 = pd.read_csv('생활이동_자치구_2022.12_02시.csv', sep=",", encoding = 'cp949')
seoul_moving_3 = pd.read_csv('생활이동_자치구_2022.12_03시.csv', sep=",", encoding = 'cp949')
seoul_moving_4 = pd.read_csv('생활이동_자치구_2022.12_04시.csv', sep=",", encoding = 'cp949')
seoul_moving_5 = pd.read_csv('생활이동_자치구_2022.12_05시.csv', sep=",", encoding = 'cp949')
seoul_moving_6 = pd.read_csv('생활이동_자치구_2022.12_06시.csv', sep=",", encoding = 'cp949')
seoul_moving_7 = pd.read_csv('생활이동_자치구_2022.12_07시.csv', sep=",", encoding = 'cp949')
seoul_moving_8 = pd.read_csv('생활이동_자치구_2022.12_08시.csv', sep=",", encoding = 'cp949')
seoul_moving_9 = pd.read_csv('생활이동_자치구_2022.12_09시.csv', sep=",", encoding = 'cp949')
seoul_moving_10 = pd.read_csv('생활이동_자치구_2022.12_10시.csv', sep=",", encoding = 'cp949')
seoul_moving_11 = pd.read_csv('생활이동_자치구_2022.12_11시.csv', sep=",", encoding = 'cp949')
seoul_moving_12 = pd.read_csv('생활이동_자치구_2022.12_12시.csv', sep=",", encoding = 'cp949')
seoul_moving_13 = pd.read_csv('생활이동_자치구_2022.12_13시.csv', sep=",", encoding = 'cp949')
seoul_moving_14 = pd.read_csv('생활이동_자치구_2022.12_14시.csv', sep=",", encoding = 'cp949')
seoul_moving_15 = pd.read_csv('생활이동_자치구_2022.12_15시.csv', sep=",", encoding = 'cp949')
seoul_moving_16 = pd.read_csv('생활이동_자치구_2022.12_16시.csv', sep=",", encoding = 'cp949')
seoul_moving_17 = pd.read_csv('생활이동_자치구_2022.12_17시.csv', sep=",", encoding = 'cp949')
seoul_moving_18 = pd.read_csv('생활이동_자치구_2022.12_18시.csv', sep=",", encoding = 'cp949')
seoul_moving_19 = pd.read_csv('생활이동_자치구_2022.12_19시.csv', sep=",", encoding = 'cp949')
seoul_moving_20 = pd.read_csv('생활이동_자치구_2022.12_20시.csv', sep=",", encoding = 'cp949')
seoul_moving_21 = pd.read_csv('생활이동_자치구_2022.12_21시.csv', sep=",", encoding = 'cp949')
seoul_moving_22 = pd.read_csv('생활이동_자치구_2022.12_22시.csv', sep=",", encoding = 'cp949')
seoul_moving_23 = pd.read_csv('생활이동_자치구_2022.12_23시.csv', sep=",", encoding = 'cp949')

total = pd.concat([seoul_moving_0, seoul_moving_1,seoul_moving_2,seoul_moving_3,seoul_moving_4,seoul_moving_5,seoul_moving_6,seoul_moving_7,seoul_moving_8,seoul_moving_9,
                  seoul_moving_10, seoul_moving_11,seoul_moving_12,seoul_moving_13,seoul_moving_14,seoul_moving_15,seoul_moving_16,seoul_moving_17,seoul_moving_18,seoul_moving_19,seoul_moving_20,
                  seoul_moving_20, seoul_moving_21,seoul_moving_22,seoul_moving_23])

In [35]:
total.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 31120983 entries, 0 to 1016071
Data columns (total 10 columns):
 #   Column       Dtype 
---  ------       ----- 
 0   대상연월         int64 
 1   요일           object
 2   도착시간         int64 
 3   출발 시군구 코드    int64 
 4   도착 시군구 코드    int64 
 5   성별           object
 6   나이           int64 
 7   이동유형         object
 8   평균 이동 시간(분)  int64 
 9   이동인구(합)      object
dtypes: int64(6), object(4)
memory usage: 2.6+ GB


In [36]:
# 결측치 처리
total.replace("*","1.5", inplace=True)

In [37]:
total['이동인구(합)'].value_counts()

1.5        4144299
6.05        426963
3.3         417221
3.29        305270
3.31        247335
            ...   
565.04           1
709.52           1
686.85           1
1812.57          1
740.17           1
Name: 이동인구(합), Length: 138998, dtype: int64

In [38]:
total = total.astype({'이동인구(합)':'float'})
total.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 31120983 entries, 0 to 1016071
Data columns (total 10 columns):
 #   Column       Dtype  
---  ------       -----  
 0   대상연월         int64  
 1   요일           object 
 2   도착시간         int64  
 3   출발 시군구 코드    int64  
 4   도착 시군구 코드    int64  
 5   성별           object 
 6   나이           int64  
 7   이동유형         object 
 8   평균 이동 시간(분)  int64  
 9   이동인구(합)      float64
dtypes: float64(1), int64(6), object(3)
memory usage: 2.6+ GB


In [40]:
# 이동유형, 요일, 성별 변수의 value수를 파악하기

In [41]:
total['이동유형'].value_counts()

EH    6364420
HE    5847032
EE    5530652
WH    3400466
WE    2993255
HW    2778912
EW    2425869
WW    1087718
HH     692659
Name: 이동유형, dtype: int64

In [42]:
total['요일'].value_counts()

금    4977733
토    4944014
목    4675529
화    4190873
수    4162571
월    4151933
일    4018330
Name: 요일, dtype: int64

In [43]:
total['성별'].value_counts()

M    17461928
F    13659055
Name: 성별, dtype: int64

In [44]:
# 서울 지역의 데이터만 골라보기

In [45]:
total = total[(total["출발 시군구 코드"] < 20000) & (total["도착 시군구 코드"] < 20000)]
total

대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분)  \
0       202212  일     0      11010      11010  F   0   EE           27   
1       202212  일     0      11010      11010  F  10   EH           75   
2       202212  일     0      11010      11010  F  10   HH           59   
3       202212  일     0      11010      11010  F  10   WW           66   
4       202212  일     0      11010      11010  F  10   EE           96   
...        ... ..   ...        ...        ... ..  ..  ...          ...   
968417  202212  토    23      11250      11250  M  80   WW           10   
968418  202212  토    23      11250      11250  M  80   WH           34   
968419  202212  토    23      11250      11250  M  80   HW           55   
968420  202212  토    23      11250      11250  M  80   EH          193   
968421  202212  토    23      11250      11250  M  80   EE          110   

        이동인구(합)  
0         83.82  
1         17.68  
2         11.80  
3          6.00  
4         11.92  
...         ...  
968417     9.52  
968418    19.03  
968419     9.46  
968420    28.29  
968421     9.46  

[11413385 rows x 10 columns]

In [46]:
# 서울지역의 구별로 groupby하여 데이터를 확인하기

In [47]:
total.groupby(by=['출발 시군구 코드'], as_index = False)['이동인구(합)'].sum().sort_values(by = '이동인구(합)', ascending = False)

출발 시군구 코드      이동인구(합)
22      11230  51962571.58
23      11240  39645897.85
21      11220  33171449.82
18      11190  28427478.60
13      11140  27134444.08
15      11160  25086814.06
24      11250  23991951.65
1       11020  23839579.77
10      11110  22351840.40
0       11010  22316269.18
20      11210  21051039.74
7       11080  21018653.00
4       11050  19774366.16
3       11040  19139157.29
12      11130  19028002.32
5       11060  18977244.76
11      11120  18222018.26
19      11200  18134282.99
16      11170  17091461.32
2       11030  16369553.71
14      11150  16026129.68
6       11070  15441943.77
8       11090  13776874.07
9       11100  11620349.59
17      11180   9345424.06

In [48]:
# '도착시군구코드'를 기준으로 각 구별로 평균 이동시간(분)의 평균과 합을 구하기

In [50]:
# '도착시군구코드'를 기준으로 각 구별로 이동인구(합)의 평균과 합을 구하기

In [51]:
total_mean = total.groupby(by=['도착 시군구 코드'], as_index = False)[['평균 이동 시간(분)','이동인구(합)']].mean()

In [52]:
total_sum = total.groupby(by=['도착 시군구 코드'], as_index = False)[['평균 이동 시간(분)','이동인구(합)']].sum()

In [61]:
total_sum.rename(columns={"평균 이동 시간(분)":'총 이동 시간', '이동인구(합)':'총 이동 인구'},inplace=True)
total_mean.rename(columns={"평균 이동 시간(분)":'평균 이동 시간', '이동인구(합)':'평균 이동 인구'},inplace=True)

In [63]:
df_total = pd.merge(total_sum,total_mean, how = 'inner', on = '도착 시군구 코드')
df_total

도착 시군구 코드   총 이동 시간      총 이동 인구   평균 이동 시간   평균 이동 인구
0       11010  31146122  21760995.87  60.855209  42.517972
1       11020  31314213  23321664.68  59.867001  44.586722
2       11030  28417572  16460904.97  59.681473  34.570549
3       11040  28843060  19037852.51  59.445836  39.237205
4       11050  28705429  19956070.26  62.553507  43.487320
5       11060  30258217  19051343.19  61.697824  38.846520
6       11070  26859016  15662353.67  65.694541  38.308594
7       11080  30274884  21194601.10  62.060698  43.446962
8       11090  26117669  13903441.47  67.427755  35.894392
9       11100  23775387  11780062.10  69.364937  34.368453
10      11110  30840466  22579997.96  69.050251  50.555479
11      11120  27352358  18485127.81  69.561528  47.010709
12      11130  28610686  19025084.57  62.499997  41.560266
13      11140  32223748  27075664.08  62.512242  52.525251
14      11150  24362877  16178614.90  64.056869  42.538137
15      11160  32199344  25495774.22  71.263974  56.427553
16      11170  28034504  17129909.86  66.870460  40.859826
17      11180  21406113   9311781.82  67.987642  29.575014
18      11190  32620231  28139534.62  62.460591  53.881040
19      11200  27974590  18342522.13  60.422760  39.618304
20      11210  29480860  21347870.30  65.121138  47.155938
21      11220  34633665  32912085.69  65.127072  61.889718
22      11230  39604247  51011034.01  66.941243  86.221613
23      11240  36704872  39608161.90  72.946099  78.716005
24      11250  29080627  24172344.02  73.217384  60.859616

In [57]:
# 어느 요일의 이동이 제일 많을까요?

In [58]:
total.groupby(by=['요일'], as_index = False)[['평균 이동 시간(분)']].sum()

요일  평균 이동 시간(분)
0  금    117902250
1  목    115273044
2  수    110169958
3  월     99520356
4  일     85488819
5  토    109151766
6  화    103334564

In [59]:
total.groupby(by=['요일'], as_index = False)[['이동인구(합)']].sum()

요일      이동인구(합)
0  금  97440817.16
1  목  94613351.82
2  수  74719502.00
3  월  74578960.96
4  일  53953085.59
5  토  80959804.08
6  화  76679276.10